src/main/java/org/zeplinko/commons/lang/ext/core/Retryable.java

```diff
@@ -0,0 +1,225 @@
+     */
+    public Retryable<R> backoff(@Nonnull Backoff strategy) {
+        this.backoffStrategy = Objects.requireNonNull(strategy);
+        return this;
+    }
+
+    /**
+     * Executes the configured task applying all retry rules and validation.
+     *
+     * @return a {@link Result} containing either the successful value or the
+     *         exception ultimately thrown/raised
+     */
+    public Result<R, Exception> run() {
+        int attempt = 0;
+        Exception lastException = null;
+
+        while (true) {
+            attempt++;
+            try {
+                R value = task.call();
+                if (resultPredicate.test(value)) {
+                    return Result.success(value);
+                }
+            } catch (Exception e) {
+                if (!exceptionPredicate.test(e)) {
+                    return Result.failure(e);
+                }
+                lastException = e;
+            }
+
+            if (attempt > maxRetries) {
+                return Result.failure(
+                        lastException != null ? lastException
+                                : new ValidationException("Validation failed after " + attempt + " attempts")
+                );
+            }
+
+            long sleepMillis = backoffStrategy.nextDelay(baseDelayMillis, attempt);
+            if (sleepMillis > 0) {
+                try {
+                    Thread.sleep(sleepMillis);
+                } catch (InterruptedException ie) {
+                    Thread.currentThread().interrupt();
+                    return Result.failure(ie);
+                }
+            }
+        }
+    }
+
+    /**
+     * Built-in strategies for calculating the delay before each retry.
+     * <ul>
+     * <li><b>FIXED</b> – always returns the base delay</li>
+     * <li><b>LINEAR</b> – {@code baseDelay * attempt}</li>
+     * <li><b>EXPONENTIAL</b> – {@code baseDelay * 2^(attempt-1)}</li>
+     * </ul>
+     */
+    public enum Backoff {
+        FIXED {
+            @Override
+            long nextDelay(long base, int attempt) {
+                return base;
+            }
+        },
+        LINEAR {
+            @Override
+            long nextDelay(long base, int attempt) {
+                return base * attempt;
+            }
+        },
+        EXPONENTIAL {
+            @Override
+            long nextDelay(long base, int attempt) {
+                if (attempt <= 0 || base <= 0)
+                    return 0;
+
+                int shift = attempt - 1;
+                if (shift >= Long.SIZE - 1)
+                    return Long.MAX_VALUE;
+
+                long multiplier = 1L << shift;
+                if (base > Long.MAX_VALUE / multiplier)
+                    return Long.MAX_VALUE;
+                return base * multiplier;
+            }
+        };
+
+        abstract long nextDelay(long base, int attempt);
+    }
+
+    private static final class ValidationException extends Exception {
```

Comment:
This exception in only case where the Success Result doesn't satisfy the criteria. Doesn't the name sound too generic here?

Reply:
Its supposed to indicate that the retry attempts has exhaused since the validation criteria was never met.

Let me see whether we can result some Custom class instead of exception here